### ISSUE TITLE
Migrate away from Bower

### ISSUE THREAD

**Issue by eddiemonge** [yeoman/yeoman#1521]
With some of Bower's [recent](https://github.com/bower/bower.json-spec/pull/43) [changes](https://github.com/bower/json/pull/25) [...](https://github.com/bower/bower/issues/1829) (off the top of my head), it seems like they are trying to be more like npm and less like something front end developers want to use. I used to be a big supporter of Bower but now I feel like its almost useless for front end dev work. If they go through with adding files to actually list the usable things, then it will be completely irrelevant when npm can be used to accomplish the same thing.

We should start looking at what it would take to use npm (or a better suited package manager) if Bower continues down this course. I still want to have faith in them but a breaking change like this without the appropriate semver (even if it was only the spec) is hard to deal with. I don't want to say that @bower owes a lot of its popularity to Yeoman but to an extent it is true. If they make it harder for us to use then that is even more sad.

I just don't know what to do in this situation. I think it comes down to wiredep and if it can easily be switched then there isn't much holding us back from a wholesale switch.

**Comment by addyosmani** [comment]
I think now is a good time to be discussing this. 

At TodoMVC, we've been switching over to [npm](https://github.com/tastejs/todomvc/issues/1110) for the majority of our applications (driven by @sindresorhus). Whilst it hasn't been entirely friction-free, it's been a lot more plausible than it was last year. I personally think improving Yeoman's support for npm (over Bower) would offer a lot of value to our users and atm, makes sense. Especially now that npm3 installs deps as [flat](https://github.com/npm/npm/blob/multi-stage/CHANGELOG.md#flat-flat-flat) as possible, unless there are conflicts.

Some of us ([Polymer](https://github.com/npm/npm/wiki/npm-%E2%87%94-Polymer-brainstorming-session), [Angular](https://github.com/npm/npm/wiki/npm%E2%87%94Angular-brainstorming-session)) have also been speaking to the npm team about their long-term front-end plans and I feel confident that even though [npm3](http://blog.npmjs.org/post/122450408965/npm-weekly-20-npm-3-is-here-ish) doesn't yet tackle everything front-enders need, we'll be at a fairly decent place by the time npm4 comes around.

Perhaps initially it would be useful to evaluate what it would take to switch over generator-webapp and generator-gulp-webapp to an npm-only workflow. This would give us a chance to bake in sufficient tooling to work around areas npm3 doesn't yet tackle for us.

Since we're talking about wiredep, cc @stephenplusplus. Might also be worth tagging https://github.com/taptapship/wiredep/issues/173.

**Comment by Gnex77** [comment]
+1

**Comment by sheerun** [comment]
Definitely npm is nowadays better suited for front-end development. Especially because of babel & webpack that among others allow for great transcompilation experience & dead code elimination. It works really great with CommonJS modules, but is pretty clumsy with bower's AMD packages.

**Comment by wookoouk** [comment]
I am glad this is being discussed, I am still a fan of Bower but due to recent talks I am looking into other options, npm3+ is getting closer to an alternative but does anyone have any other suggestions (for pre discussion research)

p.s
+1 comments are super un useful, please stop it people.

**Comment by silvenon** [comment]
Yeah, Bower is too basic and unnecessary considering we can use npm. So this is really a discussion about package consumption, right? To share my experience:

* browserify is kinda cool, but not very trivial enough to set up and requires a plugin for everything (e.g. to watch you need to use [watchify](https://www.npmjs.com/package/watchify), also, you can't use babel-core directly, but [babelify](https://www.npmjs.com/package/babelify))
* jspm is [very cool](https://youtu.be/NpMnRifyGyw) (it uses npm/GitHub as a registry). Supports all module types, CJS, AMD, ES6 etc. It makes using SystemJS a breeze, and the last time I've heard, SystemJS is the most standards-compliant and stuff?

I haven't tried webpack yet, it seemed like a steep learning curve.

**Comment by eddiemonge** [comment]
One thing that I do want to emphasize is that we have to keep in mind packages that aren't JS or aren't only JS. Some packages can be entirely styles or HTML templates or fonts or even images. Those should be accounted for as well. Forgetting about those is where most of these package managers go wrong yet they are a very important part of front end development.

**Comment by SBoudrias** [comment]
NPM allow you to publish any type of data, so that's fine on this end.

One thing I'd like to emphasize here is that moving away from bower doesn't mean we should use browserify in all our tools. There's still value in having a simpler setup.

**Comment by wookoouk** [comment]
But isn't flooding npmjs index with CSS frameworks a conflict? I don't want
to wade through front end components to find back end packages and
visa/vers.

**Comment by SBoudrias** [comment]
@wookoouk I might be wrong, but I believe this is a non issue. You won't accidentally fall on bootstrap when you search for an hashing library.

**Comment by wookoouk** [comment]
But npmjs is specifically for back end packages, search already returns so
may results that flooding all the front end components to the same place is
not great. Also how is separation from front and back end going to work. I
want back end packages in the normal place (node_modules) but I want front
end somewhere in public/ without a per dependency rule.
I may be being a little over the top in my worked but my main point is that
they should be separate concerns, frontend and backend, combining the two
can get messy and too complex to configure, but I hope I am wrong and
someone already has it figured.

**Comment by seyDoggy** [comment]
I favor progress and I've felt for a while that bower is too limited. Though I'd have concerns over those libraries who are registered with bower only and how limited our library choices would be for a period of time.

**Comment by silvenon** [comment]
This was my thought when I saw people publishing frontend packages on npm. I don’t like it either because they don’t share the same environment. But I guess npm is going to accommodate for that eventually.

@eddiemonge to my knowledge currently there is no good solution for this. If we’re looking into a future-friendly approach, I don’t think we should care about fonts and images, because they will be replaced with SVG, which only leaves CSS. Webpack and SystemJS can load CSS, but I have no idea if that’s good :smile:

**Comment by eddiemonge** [comment]
> But npmjs is specifically for back end packages
@wookoouk its actually not. Sit and stare at https://www.npmjs.com/ for a few seconds. Also [see this](http://blog.npmjs.org/post/101775448305/npm-and-front-end-packaging)

**Comment by SBoudrias** [comment]
NPM have always been very open on the type of projects that can be published there. A couple years ago, lots of C package where published on npm (and not C binding for node, straight C modules for C people). Last month, jQuery moved his plugin site/repository to npm.

npm have been for a long time a place for any kind of package. And this is good for you, this mean you need to learn less different tools, and this mean everyone can focused on improving a single packaging tool.

npm is not Node `require('foo')`. That is Node way of loading npm modules. We can load our frontend packages inside node_modules in anyway we want.

For the story, we use bower at Yelp. Bower don't offer any particular loading way, most Yeoman generator uses wiredep, but that's not necessary. What we ended up doing here is symlinking the relevant bower packages in our build folders so they're available publicly. - I'm not advocating for this solution, but just making a note that no one should care where your modules are installed. What you do care about is making them available to your app. And this is done through dependencies wiring. Two well known example of wiring tools: wiredep and node's `require`.

**Comment by arthurvr** [comment]
I'm totally :+1: for switching more generators to npm. Both Bower and npm have big problems and I dislike both, but npm is definitely better at the moment. Npm is also actively being developed, while [Bower](https://github.com/bower/bower/graphs/contributors) is kinda dying.

I think this should be discussed separately for every generator though. Polymer, for example, still prefers Bower it seems.

Just to clarify, us moving away from Bower just means to update the official generators, right? I would prefer the yeoman-generator methods specific to Bower (`installDependencies` for example) to stay.

**Comment by callumlocke** [comment]
I like the way JSPM manages/downloads/stores packages, but I don't like the way it tries to get involved in my application. It seems to be a heavily opinionated workflow/framework (including transpiling and bundling) combined with a package manager, which is too much ecosystem in one tool.

**Comment by CWSpear** [comment]
As far as `wiredep` is concerned, it has 2 main steps: get a list of dependencies in the correct dependency order and inject them (or if you're using it programmatically, just return it). It should be *much* more trivial to do that first step using `npm`. It's actually so trivial, that you may not need `wiredep`.

I still like injecting individual files for development (faster build time when developing, easy to just break cache on that one file, easier to debug out of the box, etc) as compared to always combining a la `browserify`.

**Comment by silvenon** [comment]
@callumlocke isn't jspm basically a CLI for npm + SystemJS?

**Comment by callumlocke** [comment]
@silvenon kind of, but it's more than that.. It uses npm as one of its available 'endpoints', but it also has its own [registry](https://github.com/jspm/registry), and it stores pacakges (even npm ones) in its own `jspm_packages` folder instead of `node_modules`, in a different structure. Plus it has built-in transpiling and bundling. I like it but I worry maybe it does too much.

**Comment by silvenon** [comment]
Having recently run into an issue with this, I agree that it does too much, especially with the `jspm_packages` business. I was just testing React and couldn't require it with Node because it's not in `node_modules`, even though it was installed from the npm registry. So it's pretty silly that I can't reuse modules like that :grin:

**Comment by Munter** [comment]
jspm is designed to put your modules in a folder that is inside your web root so the browser can access them during runtime loading. That makes perfect sense. On the other hand it makes little sense if you are not leveraging SystemJS to do the loading for you

**Comment by silvenon** [comment]
Yeah, that's convenient on one hand, but it sucks that you have to have modules installed twice if you want to use them on backend as well, especially because you have to be careful to match the version when necessary.

**Comment by eddiemonge** [comment]
So looks like everyone is on board with this. The logistics part is the part that seems to be in question.

**Comment by nmccready** [comment]
Keep in mind that npm3 is considered orders of magnitude slower than npm 2, https://github.com/npm/npm/issues/8826

**Comment by addyosmani** [comment]
I think it's worth us prototyping what npm support would look like for some of our more popular official generators, just to get a feel for what else is missing in the workflow and may require additional tooling. If we get pretty close to parity with the current offerings, we can do a write-up to help other generator authors make the same transition when they're ready.

**Comment by silvenon** [comment]
Lately I've been using webpack a lot and it's wonderful. It has a larger scope, so it would also replace any gulpfiles or gruntfiles, but with a much smaller configuration. This is an overview of my experience with it.

## Intro

Everything is an asset. You can `import` anything, whether it's a JS, CSS, PNG or WOFF. You configure how you're going to handle those files with loaders. You can also say how a particular asset will be handled, or based on where it's located.

## JavaScript

webpack treats JavaScript like a first-class citizen and supports many advanced features useful for SPAs like [code splitting](http://webpack.github.io/docs/code-splitting.html). You can import assets from it if you want:

```js
import kitty from '../images/kitty.jpg';

const img =  document.createElement('img');
img.src = kitty;
```

This tells webpack that you're using this asset. It can pick up URLs automatically in CSS and HTML, but because in JS you can compose URLs, you have to explicitly require it, but it's pretty simple.

## Styles

The configuration might look like this:

```js
var autoprefixer = require('autoprefixer');
// ...
{
  test: /\.s?css$/,
  loaders: ['style', 'css', 'postcss', 'sass']
}
// ...
postcss: function() {
  return [autoprefixer];
}
```

This would process all CSS or SCSS files with Sass, PostCSS, then CSS and finally it would inline those styles inside a `<style>` tag (but you can extract them as a separate file if you want).

### Vendor

Want to add [normalize.css](https://www.npmjs.com/package/normalize.css)? It's literally as easy as:

```sh
$ npm install --save normalize.css
```
```js
// index.js
import 'normalize.css';
```

## Fonts

webpack reads `url()`s in CSS and stuff like that automatically, so all you have to do is tell it what it should do with those assets:

```js
{
  test: /\.(eot|woff2|woff|ttf)$/i,
  loaders: ['url?limit=5000']
}
```

This inlines fonts smaller than 5kb, the rest of them are copied over. I find this much more intuitive than some awkward font task.

## Images

```js
{
  test: /\.(jpe?g|png|gif|svg)$/i,
  loaders: ['url?limit=10000', 'image-webpack']
}
```

This will first optimize the images with [image-webpack-loader](https://github.com/tcoopman/image-webpack-loader), then inline the ones which are smaller than 10kb.

## Other Stuff

### Asset Revisioning

The next-to-impossible task in Gulp is very easy with webpack, you don't even notice it because [file-loader](https://github.com/webpack/file-loader) does it by default. You can customize it just by passing a query string:

```js
{
  test: /\.ext$/i,
  loaders: ['url?name=hello-world-[hash].[ext]']
}
```

Which would output `hello-world-1m215kjdg843l1n.png` instead of just `1m215kjdg843l1n.png`.

## Reason Against Webpack

It has a much steeper learning curve. But I'm sure over time it can be made much more beginner-friendly.

**Comment by sheerun** [comment]
Nowadays I'm also big fan of Webpack, so +1 for migrating to npm + webpack. But I also agree with @nmccready that npm 3 is way slower then npm 2. Currently I prefer to use npm@2.

**Comment by stephenplusplus** [comment]
If anyone is interested in the wiredep + npm story, please feel free to chime in over here: https://github.com/taptapship/wiredep/issues/173#issuecomment-159456814

**Comment by joemidi** [comment]
Any thoughts on add support for [Yarn](https://github.com/yarnpkg/yarn)?

**Comment by mischah** [comment]
> Any thoughts on adding support for [Yarn](https://github.com/yarnpkg/yarn)? 

Its done and will land in the next release of yeoman-generator. See https://github.com/yeoman/generator/issues/980 and https://github.com/yeoman/generator/issues/964 🎉

**Comment by mmiszy** [comment]
Is moving away from Bower still on the horizon?

**Comment by mischah** [comment]
Hej @mmiszy,

It’s not something that is super urgent. Because as author of a Yeoman generator you can ignore Bower and just use Yarn or npm within the install phase like it is described in the [Yeoman authoring docs](http://yeoman.io/authoring/dependencies.html).

Beside that I would give a 👍 for to ditching Bower with the next major release of https://github.com/yeoman/generator/ since Bower recommends using Yarn for new projects.

**Comment by JoshuaKGoldberg** [comment]
Per #1779, we're not looking to make any big changes. Yeoman already is largely focused on npm as the package manager at this point. It's very usable as-is. So I think this issue has separately been as resolved as it needs to be. Cheers! 🎩
